src/metabase/upload.clj

```diff
@@ -304,15 +304,35 @@
 ;;;; |  Parsing values  |
 ;;;; +------------------+
 
+(import 'java.time.LocalDateTime)
+(import 'java.time.temporal.ChronoUnit)
+
+(def ^:private last-timestamp (atom nil))
+
+(defn monotonic-now
+  "Return an adjusted version of the current time, so that it is guaranteed to be monotonic at a 1 second granularity."
+  []
+  ^LocalDateTime
+  (swap! last-timestamp
+         (fn [^LocalDateTime prev-timestamp]
+           (let [now (LocalDateTime/now)]
+             (if (nil? prev-timestamp)
+               now
+               (let [next-second (.truncatedTo (.plus prev-timestamp 1 ChronoUnit/SECONDS) ChronoUnit/SECONDS)]
+                 (if (.isAfter now next-second)
```

Comment:
I would use `java-time` for all of this (e.g., `t/after?` instead of `.isAfter`, `t/plus` instead of `.plus`, etc.)

Reply:
It's documentation is kind of a hassle, but if you `:require [java-time.api :as t]` and have good auto-completion you should be able to find what you're after.

Reply:
If you initialize `last-timestamp` to `(t/local-date-time)` instead of `nil` then all of these `if`s can condense down to

```clojure
(swap! last-timestamp (fn [prev] (t/max (t/plus prev (t/seconds 1)) (t/local-date-time))))
```

(maybe add a newline or two to make it look nicer)

Reply:
> If you initialize last-timestamp to (t/local-date-time)

I initially considered and decided against this, to avoid possiblty rounding up the first time it's called. It isn't really possible and doesn't matter in the slightest though, so cool let's simplify.

> I would use java-time for all of this

I'm not allergic to interop unless a Clojure wrapper gives me cross runtime support or much higher abstracts. Not 100% convinced in this library -its largely at the same level of abstraction as the Java classes, but less performant, and as you say its docs and generated source aren't very clear. I needed to google how it expects time units for truncate.

Using `t/max` is very nice and that sells it for me. Also happy to go with the flow if we've standardized on this library in general.

Reply:
that might be a good thing to bring up at the eng guild meeting—I'm less an advocate for java-time in particular and more for standardizing on One Way for handling time (of which java-time is the latest attempt)